# Exercise 1 : Data Acquisition

## Essential Libraries

Let us begin by importing the essential Python Libraries

NumPy: Library for Numeric Computations in Python
Pandas: Library for Data Acquisition and Preparation

In [5]:
import numpy as np
import pandas as pd

# Problem 1 : Kaggle House Prices

In [6]:
csv_data = pd.read_csv("train.csv", header = None)
csv_data.head()

0           1         2            3        4       5      6         7   \
0  Id  MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape   
1   1          60        RL           65     8450    Pave    NaN       Reg   
2   2          20        RL           80     9600    Pave    NaN       Reg   
3   3          60        RL           68    11250    Pave    NaN       IR1   
4   4          70        RL           60     9550    Pave    NaN       IR1   

            8          9   ...        71      72     73           74       75  \
0  LandContour  Utilities  ...  PoolArea  PoolQC  Fence  MiscFeature  MiscVal   
1          Lvl     AllPub  ...         0     NaN    NaN          NaN        0   
2          Lvl     AllPub  ...         0     NaN    NaN          NaN        0   
3          Lvl     AllPub  ...         0     NaN    NaN          NaN        0   
4          Lvl     AllPub  ...         0     NaN    NaN          NaN        0   

       76      77        78             79         80  
0  MoSold  YrSold  SaleType  SaleCondition  SalePrice  
1       2    2008        WD         Normal     208500  
2       5    2007        WD         Normal     181500  
3       9    2008        WD         Normal     223500  
4       2    2006        WD        Abnorml     140000  

[5 rows x 81 columns]

In [7]:
print("Data Type : ", type(csv_data))
print("Data dims : ", csv_data.shape)

Data Type :  <class 'pandas.core.frame.DataFrame'>
Data dims :  (1461, 81)


b) How many observations (rows) and variables (columns) are in the above dataset? Check the "shape"

1461 rows, 81 columns

In [8]:
csv_data.dtypes

0     object
1     object
2     object
3     object
4     object
       ...  
76    object
77    object
78    object
79    object
80    object
Length: 81, dtype: object

c) What are the data types ("dtypes") -  Numeric/Categorical - of the variables (columns) in the dataset?

Object

In [9]:
csv_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1461 entries, 0 to 1460
Data columns (total 81 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       1461 non-null   object
 1   1       1461 non-null   object
 2   2       1461 non-null   object
 3   3       1202 non-null   object
 4   4       1461 non-null   object
 5   5       1461 non-null   object
 6   6       92 non-null     object
 7   7       1461 non-null   object
 8   8       1461 non-null   object
 9   9       1461 non-null   object
 10  10      1461 non-null   object
 11  11      1461 non-null   object
 12  12      1461 non-null   object
 13  13      1461 non-null   object
 14  14      1461 non-null   object
 15  15      1461 non-null   object
 16  16      1461 non-null   object
 17  17      1461 non-null   object
 18  18      1461 non-null   object
 19  19      1461 non-null   object
 20  20      1461 non-null   object
 21  21      1461 non-null   object
 22  22      1461 non-null   

d) What does the .info() method do? Use the .info() method on the imported dataset to check this out.


The info() method prints information about the DataFrame. The information contains the number of columns, column labels, column data types, memory usage, range index, and the number of cells in each column (non-null values).

In [7]:
csv_data.describe()

0     1     2     3     4     5     6     7     8       9   ...  \
count   1461  1461  1461  1202  1461  1461    92  1461  1461    1461  ...   
unique  1461    16     6   111  1074     3     3     5     5       3  ...   
top       Id    20    RL    60  7200  Pave  Grvl   Reg   Lvl  AllPub  ...   
freq       1   536  1151   143    25  1454    50   925  1311    1459  ...   

          71  72     73    74    75    76    77    78      79      80  
count   1461   8    282    55  1461  1461  1461  1461    1461    1461  
unique     9   4      5     5    22    13     6    10       7     664  
top        0  Gd  MnPrv  Shed     0     6  2009    WD  Normal  140000  
freq    1453   3    157    49  1408   253   338  1267    1198      20  

[4 rows x 81 columns]

e) What does the .describe() method do? Use the .describe() method on the imported dataset to check

The describe() method returns description of the data in the DataFrame. If the DataFrame contains numerical data, the description contains these information for each column: 
count: Total Number of Non-Empty values.
mean: Mean of the column values.
std: Standard Deviation of the column values.
min: Minimum value from the column.
25%: 25 percentile.
50%: 50 percentile.
75%: 75 percentile.
max: Maximum value from the column.

# Problem 2 : Summer Olympic 2016 medal

In [11]:
html_data = pd.read_html('http://en.wikipedia.org/wiki/2016_Summer_Olympics_medal_table')

In [12]:
print("HTML tables : ", len(html_data))

HTML tables :  8


b) How many tables are in this wikipedia page? Check "len" of the imported data/page to find this out

There are 8 tables

In [14]:
html_data[3]

Rank                   NOC  Gold  Silver  Bronze  Total
0                     1         United States    46      37      38    121
1                     2         Great Britain    27      23      17     67
2                     3                 China    26      18      26     70
3                     4                Russia    19      17      20     56
4                     5               Germany    17      10      15     42
..                  ...                   ...   ...     ...     ...    ...
82                   78               Nigeria     0       0       1      1
83                   78              Portugal     0       0       1      1
84                   78   Trinidad and Tobago     0       0       1      1
85                   78  United Arab Emirates     0       0       1      1
86  Totals (86 entries)   Totals (86 entries)   306     307     359    972

[87 rows x 6 columns]

c) Which one is the actual "2016 Summer Olympics medal table"? Explore all tables in the data to know

The 3rd indexed ones

In [15]:
summer_olympics_2016_df = html_data[3]

print(summer_olympics_2016_df)

                   Rank                   NOC  Gold  Silver  Bronze  Total
0                     1         United States    46      37      38    121
1                     2         Great Britain    27      23      17     67
2                     3                 China    26      18      26     70
3                     4                Russia    19      17      20     56
4                     5               Germany    17      10      15     42
..                  ...                   ...   ...     ...     ...    ...
82                   78               Nigeria     0       0       1      1
83                   78              Portugal     0       0       1      1
84                   78   Trinidad and Tobago     0       0       1      1
85                   78  United Arab Emirates     0       0       1      1
86  Totals (86 entries)   Totals (86 entries)   306     307     359    972

[87 rows x 6 columns]


In [17]:
top_20_countries_df = summer_olympics_2016_df.head(20)

top_20_countries_df

Rank            NOC  Gold  Silver  Bronze  Total
0     1  United States    46      37      38    121
1     2  Great Britain    27      23      17     67
2     3          China    26      18      26     70
3     4         Russia    19      17      20     56
4     5        Germany    17      10      15     42
5     6          Japan    12       8      21     41
6     7         France    10      18      14     42
7     8    South Korea     9       3       9     21
8     9          Italy     8      12       8     28
9    10      Australia     8      11      10     29
10   11    Netherlands     8       7       4     19
11   12        Hungary     8       3       4     15
12   13        Brazil*     7       6       6     19
13   14          Spain     7       4       6     17
14   15          Kenya     6       6       1     13
15   16        Jamaica     6       3       2     11
16   17        Croatia     5       3       2     10
17   18           Cuba     5       2       4     11
18   19    New Zealand     4       9       5     18
19   20         Canada     4       3      15     22

e) Extract the TOP 20 countries from the medal table, as above, and store these rows as a new DataFrame

# Bonus Problems

## A

In [19]:
data_census_income = pd.read_table('census+income/adult.data', sep = ",", header = None)

In [41]:
data_census_income.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [20]:
data_census_income.shape

(32561, 15)

Explore the dataset using .shape

-> The dimensions is 32561 rows and 15 columns

In [21]:
data_census_income.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       32561 non-null  int64 
 1   1       32561 non-null  object
 2   2       32561 non-null  int64 
 3   3       32561 non-null  object
 4   4       32561 non-null  int64 
 5   5       32561 non-null  object
 6   6       32561 non-null  object
 7   7       32561 non-null  object
 8   8       32561 non-null  object
 9   9       32561 non-null  object
 10  10      32561 non-null  int64 
 11  11      32561 non-null  int64 
 12  12      32561 non-null  int64 
 13  13      32561 non-null  object
 14  14      32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


Explore using .info()

-> the dtypes is int64 and memory usage is 3.7 MB

In [22]:
data_census_income.describe()

0             2             4             10            11  \
count  32561.000000  3.256100e+04  32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05     10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05      2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04      1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05      9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05     10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05     12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06     16.000000  99999.000000   4356.000000   

                 12  
count  32561.000000  
mean      40.437456  
std       12.347429  
min        1.000000  
25%       40.000000  
50%       40.000000  
75%       45.000000  
max       99.000000

explore the dataset using .describe()

count: Total Number of Non-Empty values.
mean: Mean of the column values.
std: Standard Deviation of the column values.
min: Minimum value from the column.
25%: 25 percentile.
50%: 50 percentile.
75%: 75 percentile.
max: Maximum value from the column.

# B

In [24]:
top_20_countries = {}
for year in range(2000, 2017, 4):
    url = f'http://en.wikipedia.org/wiki/{year}_Summer_Olympics_medal_table'
    html_data = pd.read_html(url)
    medal_table = html_data[3]
    top_20_countries[year] = medal_table.head(20)

In [27]:
for year, top_countries_df in top_20_countries.items():
    print(f"\nTop 20 countries in {year} Summer Olympics:")
    print(top_countries_df)


Top 20 countries in 2000 Summer Olympics:
   Rank         Nation  Gold  Silver  Bronze  Total
0     1  United States    37      24      32     93
1     2         Russia    32      28      29     89
2     3          China    28      16      14     58
3     4     Australia*    16      25      17     58
4     5        Germany    13      17      26     56
5     6         France    13      14      11     38
6     7          Italy    13       8      13     34
7     8    Netherlands    12       9       4     25
8     9           Cuba    11      11       7     29
9    10  Great Britain    11      10       7     28
10   11        Romania    11       6       9     26
11   12    South Korea     8      10      10     28
12   13        Hungary     8       6       3     17
13   14         Poland     6       5       3     14
14   15          Japan     5       8       5     18
15   16       Bulgaria     5       6       2     13
16   17         Greece     4       6       3     13
17   18         Swede